app/com/linkedin/drelephant/tuning/hbt/MRApplicationData.java

```diff
@@ -0,0 +1,740 @@
+      for (Function_Name value : Function_Name.values()) {
+        processForMemory(memoryHeuristics.get(value.name()), value.name());
+      }
+    }
+    else if(isOnlySpecificTypeHeuristicsFailed(Reducer.name())) {
+      logger.info(" No Reducer specific heuristic failed , hence trying to optimize memory for reducer ");
+      processForMemory(memoryHeuristics.get(Reducer.name()),Reducer.name());
+    }
+    else if(isOnlySpecificTypeHeuristicsFailed(Mapper.name())) {
+      logger.info(" No Mapper specific heuristic failed , hence trying to optimize memory for Mapper ");
+      processForMemory(memoryHeuristics.get(Mapper.name()),Mapper.name());
+    }
+  }
+
+  private boolean isOnlySpecificTypeHeuristicsFailed(String functionName){
+    for(String heuristicName : failedHeuristics.keySet()){
+      if(heuristicName.toLowerCase().contains(functionName.toLowerCase())){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public Map<String, Double> getSuggestedParameter() {
+    return this.suggestedParameter;
+  }
+
+  /**
+   * Only if the Heuristics severity is greater than 2 , then try to tune it.
+   * @param yarnAppHeuristicResult : Heuristics details for the application
+   * @return : If its valid heuristic for tuning.
+   */
+  private boolean isFailedHeuristic(AppHeuristicResult yarnAppHeuristicResult) {
+    if (validHeuristic.contains(yarnAppHeuristicResult.heuristicName)
+        && yarnAppHeuristicResult.severity.getValue() > HEURISTIC_THRESHOLD) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * This method ,based on the heuristic failure , will call respective methods
+   * for tuning the heuristic .
+   *
+   * Since Dr elephant runs on Java 1.5 , switch case is not used.
+   * Not able to set number of reducer from Azkaban and hence not optimizing for
+   * Reducer time
+   * @param yarnAppHeuristicResult : App heuristics results.
+   * Blanked Exception is caught ,since even if some unknown exception
+   * comes , system should work as it is . In the case of unknown exception
+   * previous applied parameters will be applied .
+   */
+  private void processHeuristics(AppHeuristicResult yarnAppHeuristicResult) {
+    logger.info("Fixing Heuristics " + yarnAppHeuristicResult.heuristicName);
+    try {
+      if (yarnAppHeuristicResult.heuristicName.equals(MAPPER_MEMORY_HEURISTIC)) {
+        processForMemory(yarnAppHeuristicResult, Mapper.name());
+      } else if (yarnAppHeuristicResult.heuristicName.equals(REDUCER_MEMORY_HEURISTIC)) {
+        processForMemory(yarnAppHeuristicResult, Reducer.name());
+      } else if (yarnAppHeuristicResult.heuristicName.equals(MAPPER_TIME_HEURISTIC)) {
+        processForNumberOfTask(yarnAppHeuristicResult, Mapper.name());
+      } else if (yarnAppHeuristicResult.heuristicName.equals(MAPPER_SPILL_HEURISTIC)) {
+        processForMemoryBuffer(yarnAppHeuristicResult);
+      }
+    } catch (Exception e) {
```

Comment:
As mentioned in the comment of this method , blank exception is caught , since system should work fine if there is some unknown exception while parsing file .

Reply:
Resolved